src/Microsoft.TestPlatform.Extensions.BlameDataCollector/BlameCollector.cs

```diff
@@ -94,53 +122,178 @@ internal BlameCollector(IBlameReaderWriter blameReaderWriter, IProcessDumpUtilit
             this.logger = logger;
 
             // Subscribing to events
-            this.events.TestHostLaunched += this.TestHostLaunched_Handler;
-            this.events.SessionEnd += this.SessionEnded_Handler;
+            this.events.TestHostLaunched += this.TestHostLaunchedHandler;
+            this.events.SessionEnd += this.SessionEndedHandler;
             this.events.TestCaseStart += this.EventsTestCaseStart;
             this.events.TestCaseEnd += this.EventsTestCaseEnd;
 
             if (this.configurationElement != null)
             {
                 var collectDumpNode = this.configurationElement[Constants.DumpModeKey];
-                this.processDumpEnabled = collectDumpNode != null;
-                if (this.processDumpEnabled)
+                this.collectProcessDumpCollectionBasedOnTrigger = collectDumpNode != null;
+                if (this.collectProcessDumpCollectionBasedOnTrigger)
+                {
+                    this.ValidateAndAddTriggerBasedProcessDumpParameters(collectDumpNode);
+                }
+
+                var collectHangBasedDumpNode = this.configurationElement[Constants.CollectDumpOnTestSessionHang];
+                this.collectProcessDumpOnTestHostHang = collectHangBasedDumpNode != null;
+                if (this.collectProcessDumpOnTestHostHang)
                 {
-                    this.ValidateAndAddProcessDumpParameters(collectDumpNode);
+                    this.ValidateAndAddHangBasedProcessDumpParameters(collectDumpNode);
                 }
             }
 
             this.attachmentGuid = Guid.NewGuid().ToString().Replace("-", string.Empty);
+
+            if (this.collectProcessDumpOnTestHostHang)
+            {
+                this.inactivityTimer = this.inactivityTimer ?? new InactivityTimer(this.CollectDumpAndAbortTesthost);
+                this.ResetInactivityTimer();
+            }
         }
 
-        private void ValidateAndAddProcessDumpParameters(XmlElement collectDumpNode)
+        /// <summary>
+        /// Disposes of the timer when called to prevent further calls.
+        /// Kills the other instance of procdump if launched for collecting trigger based dumps.
+        /// Starts and waits for a new proc dump process to collect a single dump and then
+        /// kills the testhost process.
+        /// </summary>
+        private void CollectDumpAndAbortTesthost()
         {
-            foreach (XmlAttribute attribute in collectDumpNode.Attributes)
+            EqtTrace.Info(string.Format(CultureInfo.CurrentUICulture, Resources.Resources.InactivityTimeout, (int)this.inactivityTimespan.TotalMinutes));
+            this.inactivityTimerAlreadyFired = true;
+
+            try
+            {
+                EqtTrace.Verbose("Calling dispose on Inactivity timer.");
+                this.inactivityTimer.Dispose();
+            }
+            catch
```

Comment:
What exception ?

Reply:
while disposing only one exception can occur, which is already disposed